### ISSUE TITLE
ArgumentException in MethodEmitter

### ISSUE THREAD

**Issue by zplan** [castleproject/Core#371]
When I update CastleCore from 4.2.1 to 4.3.0 I get an ArgumentException in the MethodEmitter. With the older version there was no exception and the code was working.

I am using Moq with CastleCore. Stacktrace of the error:

System.ArgumentException: Constant does not match the defined type..
    at System.Reflection.Emit.TypeBuilder.SetConstantValue(ModuleBuilder module, Int32 tk, Type destType, Object value)
   at System.Reflection.Emit.ParameterBuilder.SetConstant(Object defaultValue)
   at Castle.DynamicProxy.Generators.Emitters.MethodEmitter.CopyDefaultValueConstant(ParameterInfo from, ParameterBuilder to)
   at Castle.DynamicProxy.Generators.Emitters.MethodEmitter.DefineParameters(ParameterInfo[] parameters)
   at Castle.DynamicProxy.Generators.Emitters.MethodEmitter..ctor(AbstractTypeEmitter owner, String name, MethodAttributes attributes, MethodInfo methodToUseAsATemplate)
   at Castle.DynamicProxy.Generators.Emitters.AbstractTypeEmitter.CreateMethod(String name, MethodAttributes attributes, MethodInfo methodToUseAsATemplate)
   at Castle.DynamicProxy.Generators.MethodGenerator.Generate(ClassEmitter class, ProxyGenerationOptions options, INamingScope namingScope)
   at Castle.DynamicProxy.Contributors.CompositeTypeContributor.ImplementMethod(MetaMethod method, ClassEmitter class, ProxyGenerationOptions options, OverrideMethodDelegate overrideMethod)
   at Castle.DynamicProxy.Contributors.CompositeTypeContributor.Generate(ClassEmitter class, ProxyGenerationOptions options)
   at Castle.DynamicProxy.Generators.InterfaceProxyWithoutTargetGenerator.GenerateType(String typeName, Type proxyTargetType, Type[] interfaces, INamingScope namingScope)
   at Castle.DynamicProxy.Generators.InterfaceProxyWithTargetGenerator.<>c__DisplayClass6_0.<GenerateCode>b__0(String n, INamingScope s)
   at Castle.DynamicProxy.Generators.BaseProxyGenerator.ObtainProxyType(CacheKey cacheKey, Func`3 factory)
   at Castle.DynamicProxy.Generators.InterfaceProxyWithTargetGenerator.GenerateCode(Type proxyTargetType, Type[] interfaces, ProxyGenerationOptions options)
   at Castle.DynamicProxy.DefaultProxyBuilder.CreateInterfaceProxyTypeWithoutTarget(Type interfaceToProxy, Type[] additionalInterfacesToProxy, ProxyGenerationOptions options)
   at Castle.DynamicProxy.ProxyGenerator.CreateInterfaceProxyWithoutTarget(Type interfaceToProxy, Type[] additionalInterfacesToProxy, ProxyGenerationOptions options, IInterceptor[] interceptors)
   at Moq.CastleProxyFactory.CreateProxy(Type mockType, IInterceptor interceptor, Type[] interfaces, Object[] arguments) in C:\projects\moq4\Source\ProxyFactories\CastleProxyFactory.cs:line 103
   at Moq.Mock`1.InitializeInstancePexProtected() in C:\projects\moq4\Source\Mock.Generic.cs:line 295
   at Moq.PexProtector.Invoke(Action action) in C:\projects\moq4\Source\PexProtector.cs:line 57
   at Moq.Mock`1.InitializeInstance() in C:\projects\moq4\Source\Mock.Generic.cs:line 258
   at Moq.Mock`1.OnGetObject() in C:\projects\moq4\Source\Mock.Generic.cs:line 320
   at Moq.Mock`1.get_Object() in C:\projects\moq4\Source\Mock.Generic.cs:line 234

--> It looks like there was a change in the MethodEmitter.cs of castle core which leads now to this issue:
castleproject/Core@d75e5b3#diff-d8c107b19efb77444c39c422ff125610

**Comment by jonorossi** [comment]
Yep, there were changes in 4.3.0, see the changelog.
> Improve fidelity of default value replication of optional parameters to fulfill inspection of the generated proxies. This does not change the proxying behavior. (@stakx, #356)

Please provide code that reproduces the defect, until then there isn't much we can do.

**Comment by stakx** [comment]
@zplan: Along with a code example, please also let us know what version of .NET you have installed and/or which runtime you're running this code on. Once we have all that we can reopen. Thank you!

**Comment by stakx** [comment]
@zplan reported this over at https://github.com/moq/moq4/issues/623#issuecomment-396479978, apparently proxying the following type doesn't work:

```csharp
public interface IMyInterface
{
    ShutdownResponse Shutdown(long timeoutICM = 60);
}
```

Proxying the above works fine, but the following does not:

```csharp
public interface IMyInterface
{
	ShutdownResponse Shutdown([Optional, DefaultParameterValue(60)] long timeoutICM);
}
```

If `60` (an `int`) is changed to `60L` (a `long`), the error disappears.

So until we receive repro code from @zplan, my working assumption is that s/he doesn't actually define the `IMyInterface` type himself in source code, but it might be a type from a referenced assembly that has a default value with mismatched types embedded in metadata.

I'm a bit surprised that reflection doesn't just coerce the `int` to `long` in this case.

I suppose I could provide a fix against this additional corner case. Let's give this a little time, I'll think about a fix.

**Comment by stakx** [comment]
Update from @zplan in https://github.com/moq/moq4/issues/623#issuecomment-397190935:

> The real interface where I have the issue is in another assembly which was generated by a code generator.

> Debugging with the IMyInterfaceType shows "DefaultValue 60 object {long}" --> this works
Debugging with the real interface shows "DefaultValue 60 object {int}" --> this crashes, but the reflector shows that the interface contains a long. But somehow it is represented as an int during runtime and this leads to an ArgumentException

So it is as I assumed&mdash;some code generator embedded a metadata constant of a primitive type that doesn't match the parameter type exactly, and unlike the reflector tool and Roslyn / VS IntelliSense, Reflection cannot handle this type mismatch.

How hard should *DynamicProxy* try to interpret the embedded metadata constant so it fits the parameter? According to the ECMA spec, the runtime itself doesn't do anything with the parameter metadata constants. It also doesn't require them to match the parameter type. They're simply recorded for purely informative purposes so that tooling can do something interesting with them. Given that we don't know all the code generators out there that might produce type-mismatched constants (we've only considered the .NET compilers so far), the strategy of fail-early no longer seems appropriate to me. I think DynamicProxy should make a final attempt at converting the recorded default value to the parameter type, but if that too fails DynamicProxy should just go on without throwing.

**Comment by stakx** [comment]
Bugfix is ready: #375. Once that's merged, we should be able to proceed with a 4.3.1 release (#372).

**Comment by jonorossi** [comment]
Pull request merged.
